WalletWasabi.Tests/UnitTests/BlockchainAnalysis/PubKeyReuseTests.cs

```diff
@@ -0,0 +1,154 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Blockchain.Keys;
+using WalletWasabi.Tests.Helpers;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.BlockchainAnalysis
+{
+	public class PubKeyReuseTests
+	{
+		[Fact]
+		public void AddressReusePunishment()
+		{
+			// If there's reuse in input and output side, then output side didn't gain, nor lose anonymity.
+			var analyser = ServiceFactory.BlockchainAnalyzer();
+			var km = ServiceFactory.KeyManager();
+			var reuse = BitcoinFactory.HdPubKey(km);
+			var tx = BitcoinFactory.SmartTransaction(
+				9,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, BitcoinFactory.HdPubKey(km)) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, reuse) });
+
+			// Make the reused key anonymity set something smaller than 109 (which should be the final anonymity set)
+			reuse.AnonymitySet = 30;
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+
+			// It should be smaller than 30, because reuse also gets punishment.
+			Assert.True(tx.WalletOutputs.First().HdPubKey.AnonymitySet < 30);
+		}
+
+		[Fact]
+		public void AddressReuseIrrelevantInNormalSpend()
+		{
+			// In normal transactions we expose to someone that we own the inputs and the changes
+			// So we cannot test address reuse here, because anonsets would be 1 regardless of anything.
+			var analyser = ServiceFactory.BlockchainAnalyzer();
+			var km = ServiceFactory.KeyManager();
+			var key = BitcoinFactory.HdPubKey(km);
+			var tx = BitcoinFactory.SmartTransaction(
+				0,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, key), (Money.Coins(1.2m), 100, key), (Money.Coins(1.3m), 100, key), (Money.Coins(1.4m), 100, key) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, BitcoinFactory.HdPubKey(km)) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(1, x.HdPubKey.AnonymitySet));
+			Assert.Equal(1, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void InputSideAddressReuseHaveNoConsolidationPunishmentInSelfSpend()
+		{
+			var analyser = ServiceFactory.BlockchainAnalyzer();
+			var km = ServiceFactory.KeyManager();
+			var key = BitcoinFactory.HdPubKey(km);
+			var tx = BitcoinFactory.SmartTransaction(
+				0,
+				Enumerable.Empty<Money>(),
+				new[] { (Money.Coins(1.1m), 100, key), (Money.Coins(1.2m), 100, key), (Money.Coins(1.3m), 100, key), (Money.Coins(1.4m), 100, key) },
```

Comment:
Change the anonset of one input coin to 50. The output anonset should be 50 at maximum.

Reply:
Maybe it is not possible to have a different anonset for the same key under normal circumstances, right? So what I am saying is impossible.